### PR TITLE
Add FSDP custom wrap with torch 2.1

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

Adds FSDP custom wrap with torch 2.1. Code snippets have been pulled off dev and will be used for benchmarking. This PR will not be merged until release candidate stabilizes on torch side.

This patches init and wrap functions as in the style of torch 2.0.1 to allow us to specify custom process groups

# What issue(s) does this change relate to?

[GRT-2310](https://mosaicml.atlassian.net/browse/GRT-2310)

[GRT-2310]: https://mosaicml.atlassian.net/browse/GRT-2310?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[GRT-2395](https://databricks.atlassian.net/browse/GRT-2395)

### CODE REVIEW

File: composer/trainer/mosaic_fsdp.py [hunk 1]
```diff
@@ -600,6 +601,208 @@ def init_fn(
...
+
+        ### BEGIN CHANGE ###
+        recursive_wrap_kwargs['process_group_cache'] = {}
+        ### END CHANGE ###
+
+        _custom_recursive_wrap_t2p1p0(**recursive_wrap_kwargs, **root_kwargs)  # type: ignore[arg-type]
+
+    FullyShardedDataParallel._auto_wrap = _custom_auto_wrap_t2p1p0  # type: ignore
+
 elif version.parse(torch.__version__) >= version.parse('2.1.0'):
```
Comment (vchiley) [inline composer/trainer/mosaic_fsdp.py:806]:
```suggestion
elif version.parse(torch.__version__) >= version.parse('2.1.1'):
```

File: composer/trainer/mosaic_fsdp.py [hunk 2]
```diff
@@ -19,9 +19,9 @@ def patch_pytorch():
         raise NotImplementedError(f'Not supported for torch < 1.13.1')
 
     elif version.parse(torch.__version__) < version.parse('2.0.0'):
-        # FullyShardedDataParallel monkey path for torch < 2.0 ie torch == 1.13.1
+        # Monkey path for torch < 2.0 ie torch == 1.13.1
```
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:22]:
```suggestion
        # Monkey patch for torch < 2.0 ie torch == 1.13.1
```

**Review by dakinggg — APPROVED** [review]
Didn't really review the actual fsdp patch. assume you've got that covered

**Review by dakinggg — COMMENTED** [review]
it would be nice though if you indicated what was changed from the original torch code in each of the modified functions

File: composer/trainer/mosaic_fsdp_utils.py [hunk 3]
```diff
@@ -608,6 +611,297 @@ def init_fn_t2p0p1(
...
+    module: nn.Module,
+    auto_wrap_policy: Callable,
+    wrapper_cls: Callable,
+    ignored_modules: Set[nn.Module],
+    ignored_params: Set[nn.Parameter],
+    process_group_cache: Dict[Tuple[int], Any],
+    only_wrap_children: bool = False,
+    **kwargs: Any,
+) -> Tuple[nn.Module, int]:
+    """Modified version of https://github.com/pytorch/pytorch/blob/8292b03c47fd71beb23ae834971e044aef6f4d7c/torch/distributed/fsdp/_wrap_utils.py#L25.
```
[the remark this answers is not in the document]
Comment (b-chu) [reply composer/trainer/mosaic_fsdp_utils.py:624]: Also it'd be helpful to explicitly say what torch version is required (check consistency with _custom_recursive_wrap_t2p0p1 docustring).

File: composer/trainer/mosaic_fsdp_utils.py [hunk 4]
```diff
@@ -608,6 +611,297 @@ def init_fn_t2p0p1(
...
+                ],
+            )
+            recursive_wrap_kwargs['auto_wrap_policy'] = policy
+            _warn_on_overridden_mixed_precision(overridden_module_classes)
+
+        recursive_wrap_kwargs['process_group_cache'] = {}
+
+        _custom_recursive_wrap_t2p1p0(**recursive_wrap_kwargs, **root_kwargs)  # type: ignore[arg-type]
+
+    def init_fn_t2p1p0(
```
Comment (b-chu) [inline composer/trainer/mosaic_fsdp_utils.py:800]: Same as above.

File: setup.py [hunk 5]
```diff
@@ -78,7 +78,7 @@ def package_files(prefix: str, directory: str, extension: str):
     'torchmetrics>=0.10.0,<1.1',
     'torch_optimizer>=0.3.0,<0.4',
     'torchvision>=0.13.1,<0.17',
-    'torch>=1.13.1,<2.1.1',
```
Comment (b-chu) [inline setup.py:81]: Why is this change needed?

**Review by b-chu — CHANGES_REQUESTED** [review]
I didn't check your patch implementation; let me know if you want me to. Can you also add more details in the PR description of what/how exactly you're patching?

File: composer/trainer/mosaic_fsdp_utils.py [hunk 6]
```diff
@@ -608,6 +611,297 @@ def init_fn_t2p0p1(
         _register_all_state_dict_hooks(self)
 
 
+def _custom_recursive_wrap_t2p1p0(
```
[the remark this answers is not in the document]
Comment (mvpatel2000) [reply composer/trainer/mosaic_fsdp_utils.py:614]: I would rather not. I would rather copy paste the torch code and not try to refactor their code

File: composer/trainer/mosaic_fsdp_utils.py [hunk 7]
```diff
@@ -608,6 +611,297 @@ def init_fn_t2p0p1(
...
+                wrapper_cls=wrapper_cls,
+                ignored_modules=ignored_modules,
+                ignored_params=ignored_params,
+                process_group_cache=process_group_cache,
+                **kwargs,
+            )
+            setattr(module, name, wrapped_child)
+            # Keep track of how many parameters have been wrapped
+            total_wrapped_numel += num_wrapped_params
+        # decide if we need to wrap the current module,
```
[the remark this answers is not in the document]
Comment (mvpatel2000) [reply composer/trainer/mosaic_fsdp_utils.py:684]: torch comment, resolving

File: composer/trainer/mosaic_fsdp_utils.py [hunk 8]
```diff
@@ -687,6 +690,8 @@ def _custom_recursive_wrap_t2p1p0(
 
         module_kwargs = auto_wrap_policy(module=module, recurse=False, nonwrapped_numel=remainder)
         if not only_wrap_children and module_kwargs:
+            # CHANGE: We modify the original code to support custom FSDP kwargs and add
```
Comment (b-chu) [inline composer/trainer/mosaic_fsdp_utils.py:693]: Can you add a similar comment to these functions as well?

_custom_recursive_wrap_t1p13p1
custom_auto_wrap_t1p13p1
_custom_recursive_wrap_t2p0p1
init_fn_t2p0p1
_custom_auto_wrap_t2p0p1
_custom_auto_wrap_t2p1p0
init_fn_t2p1p0

**Review by b-chu — CHANGES_REQUESTED** [review]

**Review by b-chu — APPROVED** [review]
Thanks Mihir!

**Pull request merged**